src/test/util/mod.rs

```diff
@@ -83,28 +85,49 @@ impl TestClient {
         ))
         .unwrap();
 
-        let build_info = RunCommand::new("test".into(), doc! { "buildInfo":  1 }, None).unwrap();
+        let mut server_version = None;
+        let mut server_parameters = None;
 
-        let response = client
-            .execute_operation_with_session(build_info, &mut session)
-            .await
-            .unwrap();
+        if collect_server_info {
```

Comment:
I'm a little worried about the potential proliferation of environmental variables needed to run our tests. Right now, the only env var needed to invoke the tests (besides ones that need secret info like the Atlas connectivity tests and AWS auth) is MONGODB_URI, and if possible I'd like to keep it that way since it makes it much easier to run the tests locally. (I was going to touch on this in my comment above about the MONGODB_API_VERSION variable above as well.)

I don't have a good sense of which cases in the future that we'll actually need to have the API version set for and which ones we won't. Will it be a 50/50 split of our variants on whichever server version first implements it, where we run all our tests both with and without it? Will we just have one variant running a few tests with the API version set and then run the rest without, or vice-versa? Or will we run all of our tests on newer server versions with the API version set? If the only tests we'll miss running without an env var set are specific to the API version functionality, or if it's like TLS where we just run the entire test suite with or without it, I wouldn't be opposed to a new env var, but if this is going to be a requirement for most or all of our tests going forward, I'd really prefer not to require that we need to have another env var set when invoking the tests locally.

Reply:
Yeah I think these questions are worth sorting out. The [spec](https://github.com/alcaeus/specifications/blob/drivers-996-versioned-api/source/versioned-api/tests/README.rst#id2) is currently a little vague on this and only says the following: 

>Drivers MUST test against a server with the requireApiVersion parameter enabled that also requires authentication. Since API versioning options can't be specified using the connection string, drivers MUST rely on environment variables to receive the API version to use in tests.

Assuming we treat API version as a tls-like option and run the entire suite with different values for it (I think this is the plan since the server needs startup parameters), then I think we'd also need to introduce other environment variables for server version, etc. as suggested in my previous comment since we can't use `buildInfo` in those cases. Note, this means that we'd only introduce new environment variables when doing API version testing and normal testing can just the the URI as we do today so I don't think this should impact running tests locally too much.

Reply:
Trying to answer the various questions in the comments above here.

> I'm a little worried about the potential proliferation of environmental variables needed to run our tests.

The spec calls for running tests against a server with `requireApiVersion=true` to have a functional test suite and ensure that everything works as expected (for example, this uncovered a server bug with `getMore` commands simultaneously requiring and forbidding `apiVersion` options). Since the API version can only be declared as a client option and is not exposed as a URI option (see rationale in the spec), we have to use an environment variable or similar mechanism to declare an API version for tests. However, I don't expect this variant to be tested locally very often, unless someone is explicitly working on versioned API features.

> I don't have a good sense of which cases in the future that we'll actually need to have the API version set for and which ones we won't.

Just like we run tests on a server with and without authentication enabled, we'll also run tests on servers with `requireApiVersion=true`. For the time being this is only a single server version, but I expect us to test this with all server versions that support the feature. However, this always runs the entire test. The spec tests for versioned API test the driver logic for commands explicitly to ensure that API inheritance is implemented properly. These tests will change if we decide to change inheritance rules for the declared API version (e.g. allowing users to declare a different API version on a `MongoDatabase` or `MongoCollection` object). I hope that clears up the confusion as to what variants should be tested. Note that the evergreen pipeline currently does not test against a server with `requireApiVersion=true` for multiple reason: tests would be failing until the `getMore` issue explained above is fixed, and we currently can't use mongo-orchestration to start such a cluster as the `isMaster` command run by mongo-orchestration fails as it doesn't declare an API version.

> I worry about this approach where we sometimes don't get the server information, since in those cases all of `TestClient`'s methods panic when called.

Typically when creating test clients, they will not be running with `apiStrict=true`, only with `apiVersion=1`. The use-case for `collect_server_info` are client entities in the unified test runner. These are created with custom options, including a strict API version. After discussing this with @isabelatkinson I decided that the sensible thing would be to disable collection of this data if disabled. I'm open to alternatives, but most of these would go beyond the scope of this PR. One option would be to use a different implementation for client entities, another would be to make fetching buildinfo lazy and only do it when the info is needed. I'm not entirely happy with the added constructor argument, but it was the most logical choice to work around the limitation of not being able to test any strict API variants.

Reply:
To summarize the testing requirements based on our slack discussion:
- Drivers will be required to run their entire test suites with certain API versions, configured via environment variables (similar to auth)
- As part of normal test runs, the versioned API spec tests will be run and configure an API version they read from the JSON formats, no environment variable needed.

For the first point, I think the versioned API test readme needs to be updated to explicitly require drivers to do this and also to lay out specifically which API versions / options combos need to be tested in such a manner (as we discussed in slack). 

With regards to the second point, I'm okay with temporarily merging the `TestClient` changes for the sake of moving this project along, but the Rust driver team needs to follow up and figure out a more robust solution that doesn't introduce so much panic risk. 

Does this sound good to everyone?

Reply:
:+1: